hail/src/main/scala/is/hail/expr/ir/TableWriter.scala

```diff
@@ -217,56 +217,89 @@ case class PartitionNativeWriter(spec: AbstractTypedCodecSpec, partPrefix: Strin
-    ReferenceGenome.exportReferences(fs, referencesPath, typ.globalType)
+case class RVDSpecWriter(path: String, spec: RVDSpecMaker) extends MetadataWriter {
+  def annotationType: Type = TArray(TString)
+  def writeMetadata(
+    writeAnnotations: => IEmitCode,
+    cb: EmitCodeBuilder,
+    region: Value[Region]): Unit = {
+    cb += cb.emb.getFS.invoke[String, Unit]("mkDir", path)
+    writeAnnotations.consume(cb, {
+      cb._fatal("write annotations can't be missing!")
+    }, { case pc : PIndexableCode =>
+      val a = pc.memoize(cb, "filePaths")
+      val partFiles = cb.newLocal[Array[String]]("partFiles")
+      val n = cb.newLocal[Int]("n", a.loadLength())
+      val i = cb.newLocal[Int]("i", 0)
+      cb.assign(partFiles, Code.newArray[String](n))
+      cb.whileLoop(i < n, {
+        a.loadElement(cb, i).consume(cb, {
+          cb._fatal("file name can't be missing!")
+        }, { case s: PStringCode =>
+          cb += partFiles.update(i, s.loadString())
+        })
+        cb.assign(i, i + 1)
+      })
+      cb += cb.emb.getObject(spec)
+        .invoke[Array[String], AbstractRVDSpec]("apply", partFiles)
+        .invoke[FS, String, Unit]("write", cb.emb.getFS, path)
+    })
+  }
+}
 
+class TableSpecHelper(path: String, rowRelPath: String, globalRelPath: String, refRelPath: String, typ: TableType, log: Boolean) {
+  def write(fs: FS, partCounts: Array[Long]): Unit = {
     val spec = TableSpecParameters(
       FileFormat.version.rep,
       is.hail.HAIL_PRETTY_VERSION,
-      "references",
+      refRelPath,
       typ,
-      Map("globals" -> RVDComponentSpec("globals"),
-        "rows" -> RVDComponentSpec("rows"),
-        "partition_counts" -> PartitionCountsComponentSpec(partitionCounts)))
+      Map("globals" -> RVDComponentSpec(globalRelPath),
+        "rows" -> RVDComponentSpec(rowRelPath),
+        "partition_counts" -> PartitionCountsComponentSpec(partCounts)))
+
     spec.write(fs, path)
 
-    writeNativeFileReadMe(fs, path)
+    val nRows = partCounts.sum
+    if (log) info(s"wrote table with $nRows ${ plural(nRows, "row") } " +
+      s"in ${ partCounts.length } ${ plural(partCounts.length, "partition") } " +
+      s"to $path")
+  }
+}
 
-    using(fs.create(path + "/_SUCCESS"))(_ => ())
+case class TableSpecWriter(path: String, typ: TableType, rowRelPath: String, globalRelPath: String, refRelPath: String, log: Boolean) extends MetadataWriter {
+  def annotationType: Type = TArray(TInt64)
 
-    val nRows = partitionCounts.sum
-    info(s"wrote table with $nRows ${ plural(nRows, "row") } " +
-      s"in ${ partitionCounts.length } ${ plural(partitionCounts.length, "partition") } " +
-      s"to $path")
+  def writeMetadata(
+    writeAnnotations: => IEmitCode,
+    cb: EmitCodeBuilder,
+    region: Value[Region]): Unit = {
+    cb += cb.emb.getFS.invoke[String, Unit]("mkDir", path)
+    writeAnnotations.consume(cb, {
+      cb._fatal("write annotations can't be missing!")
+    }, { case pc: PIndexableCode =>
```

Comment:
You can also use the pattern
```
val pc = writeAnnotations.handle(cb, cb._fatal(...))
...
```
for this, which I think is clearer, and reduces indentation.